src/modalities/models/coca/coca_model.py

```diff
@@ -0,0 +1,176 @@
+import math
+from functools import partial
+from typing import Annotated, Dict, Tuple
+
+import torch
+from einops import repeat
+from pydantic import BaseModel, Field
+from torch import nn
+
+from modalities.models.coca.attention_pooling import AttentionPooling
+from modalities.models.coca.multi_modal_decoder import MultiModalDecoder
+from modalities.models.coca.text_decoder import TextDecoder
+from modalities.models.gpt2.gpt2_model import ActivationType, GPT2Config, WeightInitailizationConfig
+from modalities.models.model import NNModel
+from modalities.models.vision_transformer.vision_transformer_model import VisionTransformer, VisionTransformerConfig
+from modalities.nn.attention import AttentionConfig
+
+
+class TextDecoderConfig(BaseModel):
+    sample_key: str
+    prediction_key: str
+    block_size: Annotated[int, Field(ge=1)]
+    vocab_size: Annotated[int, Field(ge=1)]
+    n_layer: Annotated[int, Field(ge=1)]
+    n_head: Annotated[int, Field(ge=1)]
+    n_embd: Annotated[int, Field(ge=1)]
+    ffn_hidden: Annotated[int, Field(ge=1)]
+    dropout: Annotated[float, Field(ge=0.0)]
+    bias: bool
+    attention_config: AttentionConfig
+    activation: ActivationType
+    epsilon: Annotated[float, Field(ge=0.0)]
+
+
+class CoCaConfig(BaseModel):
+    prediction_key: str = "logits"
+    vision_embd_prediction_key: str  # same key as vision encoder
+    text_embd_prediction_key: str
+    vision_cls_prediction_key: str
+    text_cls_prediction_key: str
+    vision_encoder_config: VisionTransformerConfig
+    text_decoder_config: TextDecoderConfig
+    n_pool_head: Annotated[int, Field(ge=1)]
+    n_vision_queries: Annotated[int, Field(ge=1)]
+    bias_attn_pool: bool
+    epsilon_attn_pool: Annotated[float, Field(ge=0.0)]
+    n_shared_layer: Annotated[int, Field(ge=1)]
+    weight_init: WeightInitailizationConfig
+
+
+class CoCa(NNModel):
+    """CoCa
+
+    The Contrastive Captioner (CoCa) is an encoder-decoder model that integrates the concepts of CLIP
+    and generative models such as SimVLM by using contrastive and captioning losses for training.
+
+    Paper: `CoCa: Contrastive Captioners are Image-Text Foundation Models`
+    Link: https://arxiv.org/abs/2205.01917
+    """
+
+    def __init__(
+        self,
+        prediction_key: str,
+        vision_cls_prediction_key: str,
+        text_cls_prediction_key: str,
+        vision_embd_prediction_key: str,
+        text_embd_prediction_key: str,
+        n_vision_queries: int,
+        n_pool_head: int,
+        bias_attn_pool: bool,
+        epsilon_attn_pool: float,
+        vision_encoder_config: VisionTransformerConfig,
+        text_decoder_config: GPT2Config,

```

Comment:
I don’t like the name `text_decoder_config` very much, since it is used for both the `TextDecoder` and the `MultiModalDecoder`. Why not just `decoder_config`?

Reply:
I think `text_decoder_config` fits better because the `MultiModalDecoder` is also just a text decoder with additional cross-attention on multimodal input. Maybe it would be better to rename the `MultiModalDecoder` to `MultiModalTextDecoder`.

Reply:
https://github.com/Modalities/modalities/pull/16/commits/be9e01d2156f7fc00f681286c9f202bb77afd5ec